engine/helper/pka_helper.c

```diff
@@ -721,16 +723,127 @@ static void set_bignum(pka_bignum_t *bn, pka_operand_t *operand)
 }
 
 //
-// Synchronous PKA implementation
+// Synchronous/Asynchronous PKA implementation
 //
 
+
+struct poll_args {
+    ASYNC_JOB     *job;
+    int            fd;
+    pka_operand_t *operand;
+    ecc_point_t  **ecc_point;
+};
+
+
+void init_operand(pka_operand_t *operand,
+                  uint8_t       *buf,
+                  uint32_t       buf_len,
+                  uint8_t        big_endian)
+{
+    memset(operand, 0, sizeof(pka_operand_t));
+    memset(buf,     0, buf_len);
+    operand->buf_ptr    = buf;
+    operand->buf_len    = buf_len;
+    operand->actual_len = 0;
+    operand->big_endian = big_endian;
+}
+
+
+typedef enum { PKA_SYNC, PKA_ASYNC, PKA_ASYNC_ERROR_PIPE,
+               PKA_ASYNC_ERROR_OPERAND, PKA_ASYNC_ERROR_ARGS } pka_mode_t;
+
+typedef enum { PKA_ALGO_MOD, PKA_ALGO_ECC } pka_algo_t;
+
+
+pka_mode_t init_pka_async_job(pka_handle_t handle,     pka_algo_t    algo,
+                              pka_operand_t **operand, ecc_point_t **ecc_point)
+{
+    ASYNC_JOB *job = ASYNC_get_current_job();
+
+    if ( job ) {
+        int               pipefds[2] = {0, 0}, fd = 0;
+        struct poll_args *p_args = malloc(sizeof(struct poll_args));
+
+        uint8_t  *buf = malloc(MAX_BUF + 1);
+        size_t    numfds = 0;
+        ASYNC_WAIT_CTX *ctx = ASYNC_get_wait_ctx(job);
+
+        if (PKA_ALGO_MOD == algo )
+        {
+            *operand = malloc(sizeof (pka_operand_t));
+            if (*operand  == NULL) {
+                printf("Failed to malloc operand\n");
+                return PKA_ASYNC_ERROR_OPERAND;
+            }
+            init_operand((pka_operand_t *)*operand, buf, MAX_BUF, 0);
+        }
+        p_args = OPENSSL_malloc(sizeof(struct poll_args));
+        if (p_args == NULL) {
+            printf("%s Failed to malloc p_args\n", __FUNCTION__);
+	    free(operand);
+            return PKA_ASYNC_ERROR_ARGS;
+        }
+
+        if (pipe(pipefds) != 0) {
+            printf("%s Failed to create pipe\n", __FUNCTION__);
+	    if (PKA_ALGO_MOD == algo )
+                free(operand);
+
+            free(p_args);
+            return PKA_ASYNC_ERROR_PIPE;
+        }
+
+        if (!ASYNC_WAIT_CTX_get_all_fds(ctx, &fd, &numfds)
+                 || numfds >= 1) {
+            ASYNC_WAIT_CTX_clear_fd(ctx, (void*) handle);
+            close(fd);
+            close(fd+1);
+        }
+
+        p_args->fd = pipefds[0];
+
+        switch (algo)
+        {
+            case PKA_ALGO_MOD:
+                p_args->operand = *operand;
+                break;
+            case PKA_ALGO_ECC:
+                p_args->operand   = NULL;
+                p_args->ecc_point = ecc_point;
+                break;
+            default:
+                break;
+        }
+
+        p_args->job = job;
+
+        ASYNC_WAIT_CTX_set_wait_fd(ASYNC_get_wait_ctx(job), (void*) handle,
+                                   pipefds[0], p_args, NULL);
+        return PKA_ASYNC;
+    }
+    return PKA_SYNC;
+}
```

Comment:
do we need to free p_args?

Reply:
Yes, good catch. Shall be freed in polling_func.